sw/device/silicon_creator/lib/ownership/owner_block.c

```diff
@@ -0,0 +1,175 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/ownership/owner_block.h"
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/boot_data.h"
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "flash_ctrl_regs.h"
+
+enum {
+  kFlashBankSize = FLASH_CTRL_PARAM_REG_PAGES_PER_BANK,
+};
+
+rom_error_t owner_block_parse(const owner_block_t *block,
+                              owner_config_t *config,
+                              owner_application_keyring_t *keyring) {
+  if (block->header.tag != kTlvTagOwner)
+    return kErrorOwnershipInvalidTag;
+  if (block->header.length != sizeof(owner_block_t))
+    return kErrorOwnershipInvalidTagLength;
+
+  config->unknown_tags = 0;
+  config->sram_exec = block->sram_exec_mode;
+  config->flash = NULL;
+  config->info = NULL;
+  config->rescue = NULL;
+  uint32_t remain = sizeof(block->data);
+  uint32_t offset = 0;
+  while (remain) {
+    const tlv_header_t *item = (const tlv_header_t *)(block->data + offset);
+    if (item->tag == 0 || item->length == 0) {
+      break;
+    }
+    if (item->length > remain) {
+      return kErrorOwnershipInvalidTagLength;
+    }
+    remain -= item->length;
+    offset += item->length;
+    switch (item->tag) {
+      case kTlvTagApplicationKey:
+        if (keyring->length < ARRAYSIZE(keyring->key)) {
+          keyring->key[keyring->length++] =
+              (const owner_application_key_t *)item;
+        }
+        break;
+      case kTlvTagFlashConfig:
+        if (config->flash)
+          return kErrorOwnershipDuplicateItem;
+        config->flash = (const owner_flash_config_t *)item;
+        break;
+      case kTlvTagInfoConfig:
+        if (config->info)
+          return kErrorOwnershipDuplicateItem;
+        config->info = (const owner_flash_info_config_t *)item;
+        break;
+      case kTlvTagRescueConfig:
+        if (config->rescue)
+          return kErrorOwnershipDuplicateItem;
+        config->rescue = (const owner_rescue_config_t *)item;
+        break;
+      default:
+        /* unrecognized item. skip. */
+        config->unknown_tags += 1;
+    }
+  }
+  return kErrorOk;
+}
+
+#define MUBI(ex_) (ex_) ? kMultiBitBool4True : kMultiBitBool4False
+
+rom_error_t owner_block_flash_apply(const owner_flash_config_t *flash,
+                                    uint32_t config_side,
+                                    uint32_t primary_side) {
+  if (!flash)
```

Comment:
Is there a case where this evaluates to true in production environments? If not, It would probably be good to replace it for a hardened check. Same for other functions.

We can chat to @johannheyszl about the likelihood of faults to zero in the current implementation.

Reply:
@cfrantz and I have chatted

Reply:
I've changed the use of NULL to use `kHardenedBoolFalse`.